lib/identity_cache/cache_fetcher.rb

```diff
@@ -25,27 +69,184 @@ def fetch_multi(keys, &block)
       results
     end
 
-    def fetch(key)
-      result = nil
-      yielded = false
-      @cache_backend.cas(key) do |value|
-        yielded = true
-        unless IdentityCache::DELETED == value
-          result = value
-          break
+    def fetch(key, fill_lock_duration: nil, lock_wait_limit: 2)
+      if fill_lock_duration && IdentityCache.should_fill_cache?
+        fetch_with_fill_lock(key, fill_lock_duration, lock_wait_limit) do
+          yield
+        end
+      else
+        fetch_without_fill_lock(key) { yield }
+      end
+    end
+
+    private
+
+    def fetch_without_fill_lock(key)
+      data = nil
+      upsert(key) do |value|
+        value = nil if value == IdentityCache::DELETED || FillLock.cache_value?(value)
+        unless value.nil?
+          return value
         end
-        result = yield
+        data = yield
         break unless IdentityCache.should_fill_cache?
-        result
+        data
+      end
+      data
+    end
+
+    def fetch_with_fill_lock(key, fill_lock_duration, lock_wait_limit)
+      raise ArgumentError, 'fill_lock_duration must be greater than 0.0' unless fill_lock_duration > 0.0
+      raise ArgumentError, 'lock_wait_limit must be greater than 0' unless lock_wait_limit > 0
+      lock = nil
+      using_fallback_key = false
+      expiration_options = EMPTY_HASH
+      (lock_wait_limit + 2).times do # +2 is for first attempt and retry with fallback key
+        result = fetch_or_take_lock(key, old_lock: lock, **expiration_options)
+        case result
+        when FillLock
+          lock = result
+          if lock.client_id == client_id # have lock
+            data = begin
+              yield
+            rescue
+              mark_fill_failure_on_lock(key, expiration_options)
+              raise
+            end
+
+            if !fill_with_lock(key, data, lock, expiration_options) && !using_fallback_key
+              # fallback to storing data in the fallback key so it is available to clients waiting on the lock
+              expiration_options = fallback_key_expiration_options(fill_lock_duration)
+              @cache_backend.write(lock_fill_fallback_key(key, lock), data, **expiration_options)
+            end
+            return data
+          else
+            raise LockWaitTimeout if lock_wait_limit <= 0
```

Comment:
Is it better to fail or to yield? I suppose the worker will have waited a long time by this point, but if `lock_wait_limit * fill_lock_duration` is much less than the request timeout, falling back to the database would make sense. In Shopify configuration I imagine that wouldn't be the case though.

Reply:
It would certainly be a problem if the `lock_wait_limit * fill_lock_duration` was configured to be too low.  That's probably true of all timeouts, although this one we might only notice during concurrent reads of the same key, making the misconfiguration more subtle.

If this is configured appropriately, then a long delay seems indicative of database load.  If we yielded here, then that would likely make the problem much worse, since it would effectively be unleashing the thundering herd this is here to protect against.

Reply:
I suppose a lock wait timeout should be very rare anyway, since it would imply that several workers were unable to fill the value, but also didn’t encounter an error and mark a fill failure.

Under heavy load, does it seem right that the most likely failure outcome is query execution timeouts that cause fill failures to be marked?

Reply:
The lock wait time can end up adding latency in addition to the query itself.  For a single lock wait `fill_lock_duration` ends up being a minimum time to wait, so we don't want it much higher than the typical time the query takes.  In the worst case, a client could end up doing lock waits before the query, which could add up to `(lock_wait_limit - 1) * fill_lock_duration` to the query latency, so we probably want it shorter than the query timeout to avoid effectively doubling it.